cartographer_ros/cartographer_ros/offline_node.cc

```diff
@@ -132,101 +153,139 @@ void RunOfflineNode(const MapBuilderFactory& map_builder_factory) {
-        if (delayed_msg.isType<sensor_msgs::LaserScan>()) {
-          node.HandleLaserScanMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::LaserScan>());
-        }
-        if (delayed_msg.isType<sensor_msgs::MultiEchoLaserScan>()) {
-          node.HandleMultiEchoLaserScanMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::MultiEchoLaserScan>());
-        }
-        if (delayed_msg.isType<sensor_msgs::PointCloud2>()) {
-          node.HandlePointCloud2Message(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::PointCloud2>());
-        }
-        if (delayed_msg.isType<sensor_msgs::Imu>()) {
-          node.HandleImuMessage(trajectory_id, topic,
-                                delayed_msg.instantiate<sensor_msgs::Imu>());
-        }
-        if (delayed_msg.isType<nav_msgs::Odometry>()) {
-          node.HandleOdometryMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<nav_msgs::Odometry>());
-        }
-        if (delayed_msg.isType<sensor_msgs::NavSatFix>()) {
-          node.HandleNavSatFixMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::NavSatFix>());
-        }
-        clock.clock = delayed_msg.getTime();
-        clock_publisher.publish(clock);
+    playable_bag_multiplexer.AddPlayableBag(PlayableBag(
+        bag_filename, current_bag_index, ros::TIME_MIN, ros::TIME_MAX, kDelay,
+        [&](const rosbag::MessageInstance& msg) {
+          if (msg.isType<tf2_msgs::TFMessage>()) {
+            if (FLAGS_use_bag_transforms) {
+              const auto tf_message = msg.instantiate<tf2_msgs::TFMessage>();
+              tf_publisher.publish(tf_message);
+
+              for (const auto& transform : tf_message->transforms) {
+                try {
+                  // We need to keep 'tf_buffer' small because it becomes very
+                  // inefficient otherwise. We make sure that tf_messages are
+                  // published before any data messages, so that tf lookups
+                  // always work.
+                  tf_buffer.setTransform(transform, "unused_authority",
+                                         msg.getTopic() == kTfStaticTopic);
+                } catch (const tf2::TransformException& ex) {
+                  LOG(WARNING) << ex.what();
+                }
+              }
+            }
+            return false;
+          } else {
+            return true;
+          }
+        }));
+  }
 
-        LOG_EVERY_N(INFO, 100000)
-            << "Processed " << (delayed_msg.getTime() - begin_time).toSec()
-            << " of " << duration_in_seconds << " bag time seconds...";
+  while (playable_bag_multiplexer.IsMessageAvailable()) {
+    const auto next_msg_tuple = playable_bag_multiplexer.GetNextMessage();
+    const rosbag::MessageInstance& msg = std::get<0>(next_msg_tuple);
```

Comment:
use std::tie?

Reply:
Can't do it with a `MessageInstance`, since you empty-construct it. Has to be a reference.